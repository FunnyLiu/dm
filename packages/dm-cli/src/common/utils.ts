import { IContext, ILooseObject } from '../interfaces';
import { execSync } from 'child_process'
import log from './log';
import * as fs from 'fs';
import * as path from 'path';
import moment = require('moment');

/**
 * 根据.json文件中声明的usingComponents字段，获取要依赖的文件
 * @param {Object} ctx cli上下文
 * @param {Object} usingComponents json文件字段
 * @param {String} context 组件上下文
 */
export function getComponentsFiles (ctx: IContext, usingComponents: Record<string, string>, context: string): string[] {
  const npmDeps = ctx.npmDeps || {};
  const sassSuffix = ctx.opts.sassSuffix;
  const requires: string[] = [];
  Object.keys(usingComponents).forEach(name => {
    let componentPath = usingComponents[name];
    const fpl = componentPath[0]; // 组件路径第一个字母
    if (componentPath.startsWith('plugin://') || npmDeps[componentPath]) return; // plugin与npm包不处理
    if (fpl === '/') {
      componentPath = path.resolve(ctx.cwd, componentPath.substr(1)).replace(/\\/g, '/');
    } else if (fpl !== '.') {
      componentPath = './' + componentPath;
    }

    requires.push(`require('${componentPath}');`);
    requires.push(`require('${componentPath}.wxml');`);
    requires.push(`require('${componentPath}.json');`);
    const absComponentPath = path.resolve(context, `${componentPath}`) // 文件绝对路径，win下并不能用于require
    if (fs.existsSync(`${absComponentPath}.${sassSuffix}`)) { // .undefined
      requires.push(`require('${componentPath}.${sassSuffix}');`);
    } else if (fs.existsSync(`${absComponentPath}.wxss`)) {
      requires.push(`require('${componentPath}.wxss');`);
    }
  });
  return requires;
}

/** 检查是否符合小程序目录结构 */
export function checkEnv (ctx: IContext): void {
  ctx.log.start('检查环境');
  const appJsPath = path.resolve(ctx.cwd, 'app.js');
  const appJson = path.resolve(ctx.cwd, 'app.json');
  if (!(fs.existsSync(appJsPath) && fs.existsSync(appJson))) {
    throw `当前目录 ${ctx.cwd} 没有找到app.js或app.json文件，请先切换到小程序根目录再执行命令`;
  }
}

/** 深度合并 */
export function deepMerge (target: ILooseObject, source: ILooseObject): ILooseObject {
  if (toString.call(target) === '[object Object]' && toString.call(source) === '[object Object]') {
    for (const key in source) {
      if (!target[key]) {
        target[key] = source[key];
      } else {
        target[key] = deepMerge(target[key], source[key]);// 递归赋值
      }
    }
  } else if (toString.call(target) === '[object Array]' && toString.call(source) === '[object Array]') {
    // 数组进行合并
    target = target.concat(source);
  } else {
    target = source;
  }
  return target;
}

/**
 * 获取文件最后一次提交信息
 * @param {String} filePath 文件相对根目录路径
 */
export function getLastCommitInfo (filePath: string): {
  path: string;
  commit: string;
  author: string;
  dateStr: string;
} {
  const defaultResult = {
    path: filePath,
    commit: '',
    author: '',
    dateStr: ''
  };
  if (/^\./.test(filePath)) {
    return defaultResult;
  }
  try {
    const str1 = execSync(`git log '${filePath}'`, { encoding: 'utf8' });
    if (str1) {
      const commit = /commit\s+(.*)\n/.exec(str1) || [];
      const Author = /Author:\s+(.*)\n/.exec(str1) || [];
      const DateStr = /Date:\s+(.*)\n/.exec(str1) || [];
      const formateDate = DateStr[1] ? moment(new Date(DateStr[1])).format('YYYY-MM-DD HH:mm:ss') : '';
      return {
        path: filePath,
        commit: commit[1] || '',
        author: Author[1] || '',
        dateStr: formateDate
      };
    } else {
      return defaultResult;
    }
  } catch (err) {
    log.debug(`error in:(${filePath}),获取commit信息失败`);
    log.debug(err);
    return defaultResult;
  }
}

/** 每一行的修改信息 */
export function getBlameInfo (filePath: string, startLine = 1, endLine: number): Record<number, string> {
  const authors: Record<number, string> = {};
  try {
    const blameInfoStr = execSync(`git blame '${filePath}' --line-porcelain -L ${startLine},${endLine}`, { encoding: 'utf8' });
    if (blameInfoStr) {
      let num = startLine;
      const reg = /author\s+(.+)\n/g;
      let regResult = reg.exec(blameInfoStr);
      while (regResult) {
        authors[num++] = regResult[1];
        regResult = reg.exec(blameInfoStr);
      }
      return authors;
    } else {
      return authors;
    }
  } catch (err) {
    log.debug(`error in:(${filePath}),获取commit信息失败`);
    log.debug(err);
    return authors;
  }
}
