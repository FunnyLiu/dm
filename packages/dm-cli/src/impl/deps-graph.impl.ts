import log from '../common/log';
import * as path from 'path';

/* 顶点结构类型 */
type VexNodeType = 'npm'|'file'|'var';
/* 顶点结构类型 */
type DepType = 'require' | 'import';

/** 顶点结构 */
class VexNode {
  /** 顶点标识,确保顶点唯一性 */
  private _key: string;
  /** 顶点类型 */
  private _type: string;
  /** 入弧链表头，即进入该顶点的弧的链表头 */
  firstin?: ArcBox;
  /** 出弧链表头，即从该顶点出去的弧的链表头 */
  firstout?: ArcBox;
  constructor (key: string, type: VexNodeType) {
    this._key = key;
    this._type = type;
  }
  /** 供外部获取 key */
  get key (): string {
    return this._key;
  }
  /** 供外部获取 type */
  get type (): string {
    return this._type;
  }
}

/**
 * 文件类型顶点
 */
class FileVexNode extends VexNode {
  /** 文件绝对路径 */
  private _filepath: string;
  /** 文件名 */
  private _filename: string;
  /** 文件大小 */
  size?: number;
  /** 最后修改人 */
  author?: string;
  /** 最后修改时间 */
  date?: string;
  /** 导出方式 */
  exportType?: 'export' | 'exportDefault' | 'cjs';
  constructor (filepath: string) {
    super(`file|${filepath}`, 'file');
    this._filepath = filepath;
    this._filename = path.basename(filepath)
  }
  /** 供外部获取 filepath */
  get filepath (): string {
    return this._filepath;
  }
  /** 供外部获取 filename */
  get filename (): string {
    return this._filename;
  }
}

/**
 * 变量类型顶点,即模块导出的函数、常量、类等
 */
class VarVexNode extends VexNode {
  /** 变量名 */
  private _name: string;
  /** 变量所在模块路径 */
  private _filepath: string
  constructor (name: string, filepath: string) {
    super(`var|${name}|${filepath}`, 'var');
    this._filepath = filepath;
    this._name = name;
  }

  /** 供外部获取 name */
  get name (): string {
    return this._name;
  }
  /** 供外部获取 filepath */
  get filepath (): string {
    return this._filepath;
  }
}

/**
 * npm包类型顶点
 */
class NpmVexNode extends VexNode {
  /** 包名 */
  private _name: string
  constructor (name: string) {
    super(`npm|${name}`, 'npm');
    this._name = name;
  }
  /** 供外部获取 name */
  get name (): string {
    return this._name;
  }
}

/** 弧结构 */
class ArcBox {
  /** 弧标识,确保弧唯一性 */
  private _key: string;
  /** 弧起点在顶点列表的索引 */
  private _tailvexIndex: number;
  /** 弧终点在顶点列表的索引 */
  private _headvexIndex: number;
  /** 入弧链表头，指向终点相同的下一条弧 */
  headlink?: ArcBox;
  /** 出弧链表头，指向起点相同的下一条弧 */
  taillink?: ArcBox;
  /** 弧信息 */
  depType?: DepType; // 依赖方式
  constructor (key: string, tailvexIndex: number, headvexIndex: number, depTye?: DepType) {
    this._key = key;
    this._tailvexIndex = tailvexIndex;
    this._headvexIndex = headvexIndex;
    this.depType = depTye;
  }
  get key (): string {
    return this._key;
  }
  get tailvexIndex (): number {
    return this._tailvexIndex;
  }
  get headvexIndex (): number {
    return this._headvexIndex;
  }
}

/**
 * 十字链表，存储依赖关系图
 */
class DepsGraph {
  /** 顶点列表 */
  private vexList: VexNode[] = [];
  /** 顶点key存储 */
  private vexKeyMap: Record<string, number> = {};
  /** 弧列表 */
  private arcList: ArcBox[] = [];
  /** 弧key存储 */
  private arcKeyMap: Record<string, number> = {};

  /** 格式化输出图的依赖关系 */
  toString (): string {
    let s = '';
    this.vexList.forEach(vex => {
      s += `\n${vex.key} -> `;
      let firstout = vex.firstout;
      if (!firstout) {
        s += 'null '
      }
      while (firstout) {
        s += `${this.vexList[firstout.headvexIndex].key}、`;
        firstout = firstout.taillink;
      }
      s = s.substr(0, s.length - 1)
      s += `\n${vex.key} <- `;
      let firstin = vex.firstin;

      if (!firstin) {
        s += 'null '
      }
      while (firstin) {
        s += `${this.vexList[firstin.tailvexIndex].key}、`;
        firstin = firstin.headlink;
      }
      s = s.substr(0, s.length - 1)
    })
    return s;
  }

  /**
   * 添加顶点
   * @param vex 顶点结构
   * @returns
   */
  addVex (vex: VexNode): number {
    const vexKey = vex.key;
    let vexIndex = this.vexKeyMap[vexKey];
    if (typeof vexIndex === 'undefined') {
      this.vexList.push(vex);
      this.vexKeyMap[vexKey] = vexIndex = this.vexList.length - 1;
    }
    return vexIndex;
  }

  /**
   * 获取单个文件顶点
   * @param filepath 文件绝对路径
   */
  getFileVexNode (filepath: string): FileVexNode {
    const vexKey = `file|${filepath}`;
    const vexIndex = this.vexKeyMap[vexKey];
    return this.vexList[vexIndex] as FileVexNode;
  }

  /**
   * 获取单个npm顶点
   * @param name 包名
   */
  getNpmVexNode (name: string): NpmVexNode {
    const vexKey = `npm|${name}`;
    const vexIndex = this.vexKeyMap[vexKey];
    return this.vexList[vexIndex] as NpmVexNode;
  }

  /**
   * 获取单个npm顶点
   * @param name 变量名
   * @param filepath 变量所在模块路径
   */
  getVarVexNode (name: string, filepath: string): VarVexNode {
    const vexKey = `npm|${name}|${filepath}`;
    const vexIndex = this.vexKeyMap[vexKey];
    return this.vexList[vexIndex] as VarVexNode;
  }

  /**
   * 向图中添加一条弧
   * @param tailvex 弧尾节点，即起点
   * @param headvex 弧首节点，即终点
   * @param depType 依赖方式
   */
  addArc (tailvex: VexNode, headvex: VexNode, depType?: DepType): void {
    const tailvexIndex = this.addVex(tailvex);
    const headvexIndex = this.addVex(headvex);
    const arcKey = `${tailvexIndex}|${headvexIndex}`;
    if (typeof this.arcKeyMap[arcKey] === 'undefined') {
      const arc = new ArcBox(arcKey, tailvexIndex, headvexIndex, depType);
      this.arcList.push(arc);
      this.arcKeyMap[arcKey] = this.arcList.length - 1;

      tailvex = this.vexList[tailvexIndex]; // 始终使用列表中的顶点
      headvex = this.vexList[headvexIndex];

      arc.taillink = tailvex.firstout; // 将起点的出弧链表头保存为当前弧的出弧链表头，即起点相同
      tailvex.firstout = arc; // 将当前弧插入到起点的出弧链表头

      arc.headlink = headvex.firstin // 将终点的入弧链表头保存为当前弧的出弧链表头，即终点相同
      headvex.firstin = arc; // 将当前弧插入到终点的入弧链表头
    } else {
      log.debug('addArc 出现重复弧', tailvex.key, '->', headvex.key)
    }
  }

  /**
   * 获取所有顶点
   */
  getVexList (): VexNode[] {
    return this.vexList;
  }

  /**
   * 获取所有弧线
   */
  getArcList (): ArcBox[] {
    return this.arcList;
  }

  /**
   * 获取所有npm顶点
   */
  getAllNpmVexNodes (): NpmVexNode[] {
    return this.vexList.filter(item => item.type === 'npm') as NpmVexNode[];
  }

  /**
   * 获取所有文件顶点
   */
  getAllFileVexNodes (): FileVexNode[] {
    return this.vexList.filter(item => item.type === 'file') as FileVexNode[];
  }

  /**
   * 获取所有变量顶点
   */
  getAllVarVexNodes (): VarVexNode[] {
    return this.vexList.filter(item => item.type === 'var') as VarVexNode[];
  }

  /**
   * 获取指定顶点的依赖顶点列表 vexNode -> xx
   * @param vexNode 指定顶点
   * @param type 指定类型
   */
  getTailVexNodes<T extends FileVexNode|VarVexNode|NpmVexNode> (vexNode: VexNode, type?: VexNodeType): T[] {
    let firstout = vexNode.firstout;
    let vexNodeList = [];
    while (firstout) {
      vexNodeList.push(this.vexList[firstout.headvexIndex])
      firstout = firstout.taillink;
    }
    if (type) {
      vexNodeList = vexNodeList.filter(item => item.type === type);
    }
    return vexNodeList as T[]
  }

  /**
   * 获取依赖指定顶点的顶点列表 xx -> vexNode
   * @param vexNode 指定顶点
   * @param type 指定类型
   */
  getHeadVexNodes<T extends FileVexNode|VarVexNode|NpmVexNode> (vexNode: VexNode, type?: VexNodeType): T[] {
    let firstin = vexNode.firstin;
    let vexNodeList = [];
    while (firstin) {
      vexNodeList.push(this.vexList[firstin.tailvexIndex])
      firstin = firstin.headlink;
    }
    if (type) {
      vexNodeList = vexNodeList.filter(item => item.type === type);
    }
    return vexNodeList as T[];
  }
}

export { DepsGraph, VexNode, FileVexNode, VarVexNode, NpmVexNode }
