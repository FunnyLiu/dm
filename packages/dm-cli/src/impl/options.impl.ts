import { IOpts } from '../interfaces'
import { Command } from 'commander';
import * as path from 'path';
import log, { LogLevel } from '../common/log';

export function parseOpts (cli: Command): Partial<IOpts> {
  const opts: Partial<IOpts> = cli.opts();

  // version 默认为 function 传参时变为string
  if (typeof opts.version === 'function') {
    opts.version = undefined
  }

  // cli.name 默认为 function 传参时变为string
  if (typeof cli.name === 'function') {
    opts.name = undefined;
  }

  if (opts['output.build']) {
    opts.output = {
      build: path.resolve(opts['output.build'])
    }
    delete opts['output.build']
  } else if (opts['output.audit']) {
    opts.output = {
      build: path.resolve(opts['output.audit'])
    }
    delete opts['output.audit']
  } else if (opts['output.preview']) {
    opts.output = {
      build: path.resolve(opts['output.preview'])
    }
    delete opts['output.preview']
  } else if (opts['output.upload']) {
    opts.output = {
      build: path.resolve(opts['output.upload'])
    }
    delete opts['output.upload']
  }

  for (const key in opts) {
    // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
    // @ts-ignore
    if (typeof opts[key] === 'undefined') delete opts[key];
  }

  if (opts.debug) {
    log.setLevel(LogLevel.debug);
  }

  log.debug('opts:', opts)

  return opts
}
