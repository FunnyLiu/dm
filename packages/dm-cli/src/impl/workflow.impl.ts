/**
 * 组合模式工作流
 * @author chenxiaoqiang12
 */

import { IWorkList, IWorkNode } from '../interfaces'

export class WorkList implements IWorkList {
  workList: IWorkNode[] = [];
  add (work: IWorkNode): void {
    this.workList.push(work);
  }
  async execute (): Promise<void> {
    for (const work of this.workList) {
      await work.execute();
    }
  }
}

/**
 * 将函数处理成工作节点
 * @param fun
 */
export function workNodeDecorate (fun: Function): IWorkNode {
  const workNode: IWorkNode = {
    execute: async () => {
      await fun();
    }
  }
  return workNode
}
