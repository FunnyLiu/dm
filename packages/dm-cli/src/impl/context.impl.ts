/* eslint-disable @typescript-eslint/no-var-requires */
/* eslint-disable no-new */
import { IOpts, ILooseObject } from '../interfaces';
import { homeConfigPath } from '../processes/config';
import { AsyncSeriesHook, AsyncParallelHook } from 'tapable';
import * as utils from '../common/utils';
import * as path from 'path';
import * as os from 'os';
import * as fs from 'fs';
import log from '../common/log';
import Config = require('webpack-chain');
import webpack = require('webpack');
import gulp = require('gulp')

class AsyncSeriesDataHook extends AsyncSeriesHook {
  /** 用于存储钩子执行后的数据 */
  public data = {} as ILooseObject
}
class AsyncParallelDataHook extends AsyncParallelHook {
  /** 用于存储钩子执行后的数据 */
  public data = {} as ILooseObject
}

class Opts {
  private static opts: IOpts;
  public static getOpts (cwd: string): IOpts {
    if (this.opts) return this.opts;

    const defaultOpts: IOpts = {
      action: 'build',
      wxcli: os.type() === 'Windows_NT' ? '' : '/Applications/wechatwebdevtools.app/Contents/MacOS/cli',
      output: {
        build: path.resolve(cwd, '../dm-build'),
        preview: path.resolve(cwd, '../dm-build'),
        upload: path.resolve(cwd, '../dm-build'),
        audit: path.resolve(cwd, '../dm-audit')
      },
      npm: {
        maxRefer: 3
      },
      plugins: [
        [require.resolve('../plugins/plugin-build/index'), {}], // 内置插件-编译相关，包含各基本文件类型的loader及依赖分析插件
        [require.resolve('../plugins/plugin-audit/index'), {}], // 内置插件-审计相关
        [require.resolve('../plugins/plugin-new/index'), {}], // 内置插件-新增相关
        [require.resolve('../plugins/plugin-preview/index'), {}], // 内置插件-预览相关
        [require.resolve('../plugins/plugin-upload/index'), {}], // 内置插件-上传相关
        [require.resolve('../plugins/plugin-optimize/index'), {}] // 内置插件-优化相关，包含js、css tree-shaking
      ]
    };

    // 优先级：默认配置<全局配置<项目配置<用户配置
    [defaultOpts,
      ...[homeConfigPath, path.resolve(cwd, 'dm.config.js'), path.resolve(cwd, 'dm.config.profile.js')].map(p => fs.existsSync(p) ? require(p) : {})
    ].reduce(utils.deepMerge);

    this.opts = defaultOpts;

    return defaultOpts
  }
  public static formatOpts (opts: IOpts): IOpts {
    const { page } = opts;
    if (page) {
      opts.page = Array.isArray(page) ? page : page.split(',')
    }
    return opts;
  }
}

export class Context {
  /** 单例 */
  private static instance?: Context;
  /** 日志模块 */
  public log = log;
  /** 工具模块 */
  public utils = utils;
  /** 流程钩子函数 */
  public hooks = {
    /** build命令钩子 */
    build: {
      /** 版本检查：cli版本、node版本、操作系统等 */
      version: new AsyncSeriesDataHook(),
      /** 环境检查：命令执行环境等 */
      env: new AsyncSeriesDataHook(),
      /** 编译打包前：做一些预处理的事情 */
      beforePack: new AsyncSeriesDataHook(),
      /** 编译打包：TODO 替换pack流程 */
      pack: new AsyncSeriesDataHook(),
      /** 编译打包后：并行钩子，做一些拷贝、文件修改类动作 */
      afterPack: new AsyncParallelDataHook(),
      /** 发布：压缩类处理 */
      release: new AsyncSeriesDataHook(),
      /** 流程结束 */
      done: new AsyncSeriesDataHook()
    },
    /** new命令钩子 */
    new: {
      beforeRun: new AsyncSeriesDataHook(),
      run: new AsyncSeriesDataHook(),
      done: new AsyncSeriesDataHook()
    },
    /** audit 命令钩子 */
    audit: {
      beforeRun: new AsyncSeriesDataHook(),
      run: new AsyncSeriesDataHook(),
      done: new AsyncSeriesDataHook()
    },
    /** preview 命令钩子 */
    preview: {
      beforeRun: new AsyncSeriesDataHook(),
      run: new AsyncSeriesDataHook(),
      done: new AsyncSeriesDataHook()
    },
    /** uoload 命令钩子 */
    upload: {
      beforeRun: new AsyncSeriesDataHook(),
      run: new AsyncSeriesDataHook(),
      done: new AsyncSeriesDataHook()
    }
  }
  /** gulp api */
  public gulp = gulp;
  /** webpack api */
  public webpack = webpack;
  /** webpack-chain */
  public config = new Config();
  /** 当前工作目录，/User/xxx/app */
  public cwd: string;
  /** 配置项 */
  public opts: IOpts;
  /** 目标目录 */
  public dist: string;
  /** 当前项目npm依赖，即根目录下 package.json 中的 dependencies 字段 */
  public npmDeps: Record<string, string>;
  /** 临时数据 */
  private tempData: ILooseObject ={};

  private constructor (opts: IOpts) {
    this.cwd = opts.cwd ? path.resolve(opts.cwd) : process.cwd();
    this.opts = this.mergeOpts(opts);
    this.dist = this.getDist();
    this.npmDeps = this.getNpmDeps();
    Context.getPlugins(this.cwd, (Clazz, opts) => {
      new Clazz(this, opts)
    });
  }

  public static getInstance (opts?: IOpts): Context {
    if (!this.instance) {
      if (!opts) throw new Error('首次初始化Context没有传递参数')
      this.instance = new Context(opts);
    }
    return this.instance;
  }
  /** 连续调用api时可能会用到 */
  public static destoryInstance (): void {
    this.instance = undefined;
  }

  /** 获取插件，用于提前获取插件中的cmdHooks */
  public static getPlugins (cwd: string, callback: (Clazz: any, params: ILooseObject) => void): void {
    const opts = Opts.getOpts(cwd)
    try {
      opts.plugins!.forEach(plugin => {
        if (Array.isArray(plugin)) {
          const [pluginPath, params = {}] = plugin;
          const Plugin = require(pluginPath) // TODO 配置文件中的npm包目前需要使用require.resolve('mod')转换
          callback(Plugin, params)
        } else {
          const Plugin = require(plugin)
          callback(Plugin, {})
        }
      })
    } catch (error) {
      log.error('初始化配置出错');
      log.error(error);
      process.exit(1)
    }
  }

  /** 暂存临时数据 */
  public setTempData<T> (key: string, value: T): void {
    this.tempData[key] = value
  }
  /** 获取临时数据 */
  public getTempData<T> (key: string): T {
    return this.tempData[key]
  }
  /** 优先级：默认配置<全局配置<项目配置<用户配置<命令行配置 */
  private mergeOpts (opts: IOpts): IOpts {
    // 默认配置
    const defaultOpts = Opts.getOpts(this.cwd);
    utils.deepMerge(defaultOpts, opts);

    Opts.formatOpts(defaultOpts);
    return defaultOpts
  }

  private getDist (): string {
    let distPath = this.opts.output![this.opts.action] || ''; // this.opts.output有默认值
    if (this.opts.app) {
      distPath += `-${this.opts.app}`;
    }
    if (this.opts.type) {
      distPath += `-${this.opts.type}`;
    }
    return distPath;
  }

  /** 获取项目npm依赖 */
  private getNpmDeps (): Record<string, string> {
    const npmDeps: Record<string, string> = {}
    const prjPackageJSONPath = path.resolve(this.cwd, 'package.json');
    if (!fs.existsSync(prjPackageJSONPath)) {
      return npmDeps;
    }

    const prjPackageJSON = JSON.parse(fs.readFileSync(prjPackageJSONPath, 'utf8'));
    const dependencies = prjPackageJSON.dependencies;
    if (!dependencies || Object.keys(dependencies).length === 0) {
      return npmDeps;
    } else {
      return dependencies
    }
  }
}
