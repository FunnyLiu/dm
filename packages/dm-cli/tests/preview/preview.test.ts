import * as cp from 'child_process';
import * as fs from 'fs';
import { IOpts, build, preview } from '../../dist/index';
import { Context } from '../../dist/impl/context.impl';
import log from '../../dist/common/log';
import { mockApp } from '../__mocks__';
import * as utils from '../utils';
import inquirer = require('inquirer');
process.exit = jest.fn() as never;

// 前置 mock
jest.mock('child_process');
jest.mock('inquirer');
jest.mock('../../dist/processes/build');
jest.mock('../../dist/common/log');
process.cwd = jest.fn(() => mockApp.path);

// 命令参数
const opts: IOpts = { action: 'preview' };

// 初始化 context
beforeAll(() => {
  Context.getInstance(opts);
})

afterAll(() => {
  const ctx = Context.getInstance(opts);
  const distPath = ctx.dist;
  // 删除dist目录
  utils.delDirSync(distPath);
})

// 每个用例执行后，终止 log
afterEach(log.done);

describe('dm preview', () => {
  const spawnParams = '';
  beforeEach(() => {
    // mock

    (cp.exec as unknown as jest.Mock).mockImplementationOnce((_1, callback) => callback(null, { stdout: '', stderr: '' }));

    (build as unknown as jest.Mock).mockImplementationOnce(() => {
      const ctx = Context.getInstance(opts);
      const distPath = ctx.dist;
      // 创建dist目录
      if (!fs.existsSync(distPath)) fs.mkdirSync(distPath);

      return Promise.resolve();
    });
  })

  describe('无dist目录', () => {
    beforeEach(() => {
      const ctx = Context.getInstance(opts);
      const distPath = ctx.dist;
      // 删除dist目录
      utils.delDirSync(distPath);
    })

    afterAll(() => {
      mockApp.resumefiles([`${mockApp.path}/project.config.json`]);
    })

    test('源项目已存在 project.config.json 文件', async () => {
      await preview();

      expect(spawnParams).toMatchSnapshot();
    });

    test('源项目无 project.config.json 文件', async () => {
      // 删除工具配置文件
      log.log(fs.existsSync(`${mockApp.path}/project.config.json`));
      mockApp.dropfiles([`${mockApp.path}/project.config.json`]);
      log.log(fs.existsSync(`${mockApp.path}/project.config.json`));

      await preview();

      expect(spawnParams).toMatchSnapshot();
    });
  })

  describe('有dist目录', () => {
    beforeEach(() => {
      const ctx = Context.getInstance(opts);
      const distPath = ctx.dist;
      // 删除dist目录
      if (!fs.existsSync(distPath)) fs.mkdirSync(distPath);
    })

    test('允许重新打包并覆盖原dist目录', async () => {
      // 模拟询问答案 true
      (inquirer.prompt as unknown as jest.Mock).mockImplementationOnce(() => Promise.resolve({ flag: true }));

      await preview();

      expect(spawnParams).toMatchSnapshot();
    });

    test('拒绝重新打包并覆盖原dist目录', async () => {
      // 模拟询问答案 true
      (inquirer.prompt as unknown as jest.Mock).mockImplementationOnce(() => Promise.resolve({ flag: false }));

      await preview();

      expect(spawnParams).toMatchSnapshot();
    });
  });
});
