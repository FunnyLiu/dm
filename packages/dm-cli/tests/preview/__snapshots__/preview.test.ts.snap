// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dm preview 无dist目录 源项目已存在 project.config.json 文件 1`] = `""`;

exports[`dm preview 无dist目录 源项目无 project.config.json 文件 1`] = `""`;

exports[`dm preview 有dist目录 允许重新打包并覆盖原dist目录 1`] = `""`;

exports[`dm preview 有dist目录 拒绝重新打包并覆盖原dist目录 1`] = `""`;
